lib/Ryu/Async.pm

```diff
@@ -439,7 +439,16 @@ sub udp_server {
         service  => $uri->port // 0,
         socktype => 'dgram'
     )->then(sub {
-        Future->done($server->write_handle->sockport)
+        Future->done($server->write_handle->sockport);
+    })->on_fail(sub {
+        my $err       = shift;
+        my $exception = Ryu::Exception->new(
+                type    => 'udp',
+                message => "UDP server failed to bind to port " . $uri->port // 0,
+                details => [$err]);
+        $log->errorf("%s - %s", $exception->message, $err);
+        Future->fail($exception);
+        $exception->throw;
```

Comment:
@alan-fs main problem with current implementation that we don't pass socket creation error from `socket` future to our `incoming`  stream. like we do for example  with `on_recv_error` hook at line 431.  Basically you need to fail stream if you fail to create  socket. And add handling failing of stream to your udp proxy script. @tom-binary @zakame please correct me if i'm wrong.

Reply:
When it fails here (at the beginning of binding to port) in fact no stream has been created at all to pass creation error to it. So I guess there is no advantage to write  `$src->fail` instead of `Future->fail`.

Reply:
However it's okay to remove `Exception` totally here and instead using `$udp_srv->port->on_fail` in the `udp_proxy` script to alert the initial error in port binding.